src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,731 @@
+		}
+
+		force_inline bool starts_with (const char* s) const noexcept
+		{
+			if (s == nullptr)
+				return false;
+
+			return starts_with (s, strlen (s));
+		}
+
+		force_inline void set_length_after_direct_write (size_t new_length) noexcept
+		{
+			set_index (new_length);
+			terminate ();
+		}
+
+		force_inline void set_at (size_t index, const TChar ch) noexcept
+		{
+			if (index >= buffer.size ())
+				return;
+
+			TChar *p = buffer + index;
+			if (*p == NUL) {
+				return;
+			}
+
+			*p = ch;
+		}
+
+		force_inline const TChar get_at (size_t index) const noexcept
+		{
+			ensure_valid_index (index);
+			return buffer.get () + index;
+		}
+
+		force_inline TChar& get_at (size_t index) noexcept
+		{
+			ensure_valid_index (index);
+			return *(buffer.get () + index);
+		}
+
+		force_inline const TChar* get () const noexcept
+		{
+			return buffer.get ();
+		}
+
+		force_inline TChar* get () noexcept
+		{
+			return buffer.get ();
+		}
+
+		force_inline size_t size () const noexcept
+		{
+			return buffer.size ();
+		}
+
+		char operator[] (size_t index) const noexcept
+		{
+			return get_at (index);
+		}
+
+		char& operator[] (size_t index) noexcept
+		{
+			return get_at (index);
+		}
+
+	protected:
+		template<typename Integer>
+		force_inline void append_integer (Integer i) noexcept
+		{
+			static_assert (std::is_integral_v<Integer>);
+
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			if constexpr (BoundsCheck) {
+				ensure_enough_space (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			}
+
+			if (i == 0) {
+				constexpr char zero[] = "0";
+				constexpr size_t zero_len = sizeof(zero) - 1;
+
+				append (zero, zero_len);
+				return;
+			}
+
+			TChar temp_buf[MAX_INTEGER_DIGIT_COUNT_BASE10 + 1];
+			TChar *p = temp_buf + MAX_INTEGER_DIGIT_COUNT_BASE10;
+			*p = NUL;
+			TChar *end = p;
+
+			uint32_t x;
+			if constexpr (sizeof(Integer) > 4) {
+				uint64_t y;
+
+				if constexpr (std::is_signed_v<Integer>) {
+					y = static_cast<uint64_t>(i > 0 ? i : -i);
+				} else {
+					y = static_cast<uint64_t>(i);
+				}
+				while (y > std::numeric_limits<uint32_t>::max ()) {
+					*--p = (y % 10) + ZERO;
+					y /= 10;
+				}
+				x = static_cast<uint32_t>(y);
+			} else {
+				if constexpr (std::is_signed_v<Integer>) {
+					x = static_cast<uint32_t>(i > 0 ? i : -i);
+				} else {
+					x = static_cast<uint32_t>(i);
+				}
+			}
+
+			while (x > 0) {
+				*--p = (x % 10) + ZERO;
+				x /= 10;
+			}
+
+			if constexpr (std::is_signed_v<Integer>) {
+				if (i < 0)
+					*--p = '-';
+			}
+
+			append (p, static_cast<size_t>(end - p));
+		}
+
+		force_inline void ensure_valid_index (size_t access_index) const noexcept
+		{
+			if (access_index < index ()) {
+				return;
+			}
+
+			log_fatal (
+				LOG_DEFAULT,
+				"Index %u is out of range (0 - %u)",
+				access_index, index ()
+			);
+			exit (1);
+		}
+
+		force_inline void ensure_enough_space (size_t length) noexcept
+		{
+			if (index () + length > buffer.size ()) {
```

Comment:
`ensure_enough_space()` is not consistent with `resize_if_needed()`: the latter has an additional `+1` in the calculation, for the terminating null.

Reply:
All addition should use `ADD_WITH_OVERFLOW_CHECK()`.

Reply:
Reviewing this PR, `ensure_enough_space(42)`/`resize_if_needed(42)` are not doing me any favors, in that it isn't explicit if `42` is the *total* size required, or the *additional* size required.  Perhaps `ensure_have_extra(42)`/`resize_for_extra(42)` would be clearer?